File: Implementation/Graph/adjacency_list.py
class Graph:
    def __init__(self) -> None:
        self.size = 0
        self.graph = {}
    
    def add_node(self, node) -> None:
        if node in self.graph:
            print('Node already exists')
            return
        self.graph[node] = []
    
    def add_edge(self, start, end) -> None:
        if start not in self.graph or end not in self.graph:
            print('Node does not exists')
            return
        for node in self.graph[start]:
            if end in node:
                print('Edge already exists')
                return
        self.graph[start].append((end, 1))
        self.size += 1
    
    def add_weighted_edge(self, start, end, weight) -> None:
        if start not in self.graph or end not in self.graph:
            print('Node does not exists')
        for node in self.graph[start]:
            if end in node: 
                print('Edge already exists')
                return
        self.graph[start].append((end, weight))
        self.size += 1
    
g1 = Graph()
g1.add_node('A')
g1.add_node('B')
g1.add_node('C')
g1.add_node('D')
g1.add_edge('A', 'B')
g1.add_edge('B', 'A')
g1.add_edge('A', 'C')
g1.add_edge('C', 'A')
g1.add_edge('B', 'C')
g1.add_edge('C', 'B')
g1.add_edge('B', 'D')
g1.add_edge('D', 'B')
print(g1.graph)

File: Implementation/Graph/go/adjacency_list.go
package main

import "fmt"

type Node struct {
	name   string
	weight int
}

type AdjacencyList struct {
	graph map[string][]Node
	size  int
}

func (al *AdjacencyList) Init() {
	al.graph = make(map[string][]Node)
}

func (al *AdjacencyList) AddNode(node string) {
	_, exists := al.graph[node]
	if exists {
		fmt.Println("Node already exists")
		return
	}
	al.graph[node] = make([]Node, 0)
}

func (al *AdjacencyList) AddEdge(from string, to string, weight int) {
	fromNode, existsFrom := al.graph[from]
	_, existsTo := al.graph[to]

	if !existsFrom || !existsTo {
		fmt.Println("Node does not exists")
	}

	for _, node := range fromNode {
		if to == node.name {
			fmt.Println("Edge already exsits")
		}
	}

	al.graph[from] = append(fromNode, Node{to, weight})
}

func main() {
	g := AdjacencyList{}
	g.Init()
	g.AddNode("A")
	g.AddNode("B")
	g.AddNode("C")
	g.AddNode("D")
	g.AddEdge("A", "B", 1)
	g.AddEdge("B", "A", 1)
	g.AddEdge("A", "C", 1)
	g.AddEdge("C", "A", 1)
	g.AddEdge("B", "C", 1)
	g.AddEdge("C", "B", 1)
	g.AddEdge("B", "D", 1)
	g.AddEdge("D", "B", 1)

	for _, node := range g.graph {
		fmt.Println(node)
	}
}
